prow/plugins/invalidcommitmsg/invalidcommitmsg.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package invalidcommitmsg adds the "do-not-merge/invalid-commit-message"
+// label on PRs containing commit messages with @mentions or
+// keywords that can automatically close issues.
+package invalidcommitmsg
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName            = "invalidcommitmsg"
+	invalidCommitMsgLabel = "do-not-merge/invalid-commit-message"
+	commentBody           = `[Keywords](https://help.github.com/articles/closing-issues-using-keywords) which can automatically close issues and at(@) mentions are not allowed in commit messages.
+
+Please remove these keywords from the following commit messages: `
+)
+
+var (
+	closeIssueRegex = regexp.MustCompile(`(([cC]los(?:e[sd]?))|([fF]ix(?:(es|ed)?))|([rR]esolv(?:e[sd]?)))[\s:]+(\w+/\w+)?#(\d+)`)
+	atMentionRegex  = regexp.MustCompile(`@[-\w]+`)
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+			Description: "The invalidcommitmsg plugin applies the '" + invalidCommitMsgLabel + "' label to pull requests whose commit messages contain @ mentions or keywords which can automatically close issues.",
+		},
+		nil
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	CreateComment(owner, repo string, number int, comment string) error
+	ListPullRequestCommits(org, repo string, number int) ([]github.RepositoryCommit, error)
```

Comment:
Do we need anything specifically from GitHub here? Why not just use the `git` client?